### ISSUE TITLE
Consider specifying how browsers tell servers they accept AST files

### ISSUE THREAD

**Issue by aickin** [syg/ecmascript-binary-ast#5]
Thanks for such a cool proposal!

Continuing on from [this thread](https://twitter.com/djeric3/status/887470011224170496), I wanted to suggest that there should be some sort of structured way for browsers to tell servers whether or not the browser understands Binary AST files, and, if supported, which version(s) are understood.

I think the most likely candidate for implementation is the HTTP Accept header, although it gets a bit complicated when combined with features like server push or inlined scripts.

If browsers don't send an Accept header or something like it, servers will have to use user-agent sniffing to figure out whether to send Binary AST or JavaScript. In my experience with sending different versions of JS to browsers, this is a pretty cumbersome and error-prone solution.

Thanks again for this!

**Comment by Qix-** [comment]
Why would `Accept` headers affect server push? Just curious. Accept can be choosy with things:

```
Accept: application/x-javascript-ast; q=0.9, application/javascript; q=0.8, text/javascript; q=0.5
```

**Comment by aickin** [comment]
@Qix- Great question; I haven't thought about it deeply, but I'll sketch out my rough ideas.

So, there are a couple of ways to use `Accept`. The simplest way would be to just add a new MIME type (let's use your suggestion `application/x-javascript-ast`) on all HTTP requests for scripts. A regular use of binary AST files would then look like this:

1. User types in `www.sitewithastscripts.com` and hits return.
2. Browser requests the HTML page with `Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8`.
3. Server returns a page with `<script src="/some_script"></script>` in it.
4. Browser requests `/some_script` with `Accept: application/x-javascript-ast,*/*`.
5. Server sees that the browser supports AST files and sends an AST version of `some_script`. If it hadn't seen that `Accept` header, it would have sent a JS version of `some_script`.

This totally works! But there are a few downsides:

1. It requires varying the content based on the Accept header, which may be difficult to support on all CDNs.
2. If you want to server push the script, the server would need to do so during step 3 before it receives the `Accept` header for the script, so the server wouldn't know whether or not the browser supports AST files.
3. If you want to inline the script in a data URL, you have the same problem as server push; the server hasn't seen the script `Accept` header yet.

So basically, to cover these use cases, I think the `Accept` header would need to specify AST file support in the original HTML request. However, that seems a little unusual, and there may be good reasons why it's not done right now. That's about the limit of my knowledge.

**Comment by Qix-** [comment]
@aickin how do servers handle this with pre-compressed assets? E.g. a server that stores both a brotli and a gzip compressed asset and serves it based on the `Accept-Encoding` header? Doing brotli compression on the fly is slow and taxing, so many servers would like to pre-compress those assets since disk space is cheap, right?

Also, correct me if I'm wrong but an agent doesn't have to support encoding (compression) on all requests, and you can't be certain they do for all subsequent requests (e.g. the agent might support brotli on the initial HTML, but not on images). I get that this is a rare case but it's still standards-compliant (again, correct me if I'm wrong).

So that's a case where server push would have to predict what the browser is going to accept and just correct itself if it doesn't, right? In which case we could apply the same logic with ASTs and the `Accept` header - assume the browser _does_ support ASTs (or doesn't, it's up to the implementation/configuration of the server) and push those first, and correct itself if the agent doesn't support it. Otherwise, the agent just disregards the push according to HTTP/2 IIRC.

This approach means that AST + Server Push only gets more optimized as adoption increases, which to me seems acceptable.

**Comment by aickin** [comment]
Great point that `Accept-Encoding` has a similar set of issues!

> an agent doesn't have to support encoding (compression) on all requests, and you can't be certain they do for all subsequent requests (e.g. the agent might support brotli on the initial HTML, but not on images)... it's still standards-compliant (again, correct me if I'm wrong).

I think that's correct, but I believe that in practice, unlike with `Accept`, all major browsers currently send the same `Accept-Encoding` header on all requests. So when you server push a compressed sub-resource, you push down the `Accept-Encoding` request header that you received on the first request. If the browser decides for some reason to change to a different `Accept-Encoding` for the sub-resource (which, again, I don't think any browsers do), then it will just ignore the server push and make a new request with its new `Accept-Encoding` header.

> In which case we could apply the same logic with ASTs and the Accept header - assume the browser does support ASTs (or doesn't, it's up to the implementation/configuration of the server) and push those first, and correct itself if the agent doesn't support it. Otherwise, the agent just disregards the push according to HTTP/2 IIRC.

Yep, you could use this same strategy for MIME type negotiation, but while it would be functionally correct, I think the perf profile for it would be pretty bad in practice. When the browser does support ASTs, you get a 90% speedup in parsing, but when it doesn't, you approximately double the amount of JavaScript sent over the wire and add a round trip. In practice, that'd be a hard tradeoff to accept, and I would suspect most engineers would fall back to user agent sniffing to avoid it.

It's also worth noting I don't think any of this helps with use case 3, when you want to inline script directly into the HTML.

> I also wonder how the folks over at WebASM are handling this, since I'm sure a broad use-case is writing some compiled code with a javascript fallback

Interesting question, and I think you're right that choosing between WASM and a fallback asm.js implementation is probably relatively common for WASM sites. I'm pretty sure that WASM is always [loaded from JavaScript on the client](https://developer.mozilla.org/en-US/docs/WebAssembly/Loading_and_running), so you'd have to do a runtime feature detect to decide whether you wanted to load WASM or JS. As a result, WASM can't be performantly server pushed or inlined without UA sniffing, I think. I'll keep my eyes peeled to see if anyone is solving it at the HTTP level.

**Comment by vigneshshanmugam** [comment]
I am also interested in how this different format would give hints to preloading scripts. There are already two potential solutions discussed for preloading modules https://github.com/whatwg/fetch/issues/486

**Comment by kannanvijayan** [comment]
@aickin The specifics of the delivery mechanism still need to be banged out, so thanks for the question.

There are a few different cases to cover, and addressing them in the standard incrementally over time may be a path to follow.  The context of script parsing can be one of:

1. Script file loaded from <script> tag.
2. Script file for a worker thread, loaded via the worker API.
3. ES6 modules.
4. Script passed to eval.
5. Am I forgetting anything?

My instinct is to quickly find a solution for the first case, and then address the others in order.  For script tags, the most straighforward, least developer-burdensome way seems to just add another attribute..

```
<script src="foo.js" binsrc="foo.binjs">
```

This requires no server configuration, is compatible with older browsers, and is easy to experiment with for developers.

For the other cases, such as workers and modules, maybe a header based approach would be better suited.  I'm not _that_ familiar with HTTP header spec details, so it's nice to read the commentary on this thread.

Thanks.

**Comment by ljharb** [comment]
I really really like the idea of allowing both; `<img src lowsrc>` was one of my favorite features back in the day. It would make it super easy to add the binary AST without compromising any aspect of the page in browsers that didn't support it.

**Comment by aickin** [comment]
@kannanvijayan Thanks for the response!

> Am I forgetting anything?

It's a good list, including several cases I hadn't thought of. I'd add:

5. HTTP/2 server push
6. `<link rel="preload" as="script">`
7. `Link:` preload HTTP header
8. Inlining into script tag, either via data URI or as content between `<script>` and `</script>`
9. `<script>` tags in HTML imports (I think these will just be the same as other `<script>` tags, but I'd want to consult an expert to make sure)

> For script tags, the most straighforward, least developer-burdensome way seems to just add another attribute..

That's a really neat idea!

Given that WebAssembly and ES6 modules may also need similar functionality, I could also imagine it being implemented more like `<picture>` or like `img`'s `srcset` attribute, where many different representations can be added. One downside is that I think this doesn't solve the push or inlining case, but the analogy of images makes me wonder if we should look at extending the Client Hints spec rather than use the `Accept` header.

**Comment by ljharb** [comment]
Versioning ECMAScript could be highly problematic; there's a reason browsers have consistently refused to do it. I don't have links off hand, but it shouldn't be too hard to locate.

**Comment by aickin** [comment]
@ljharb I haven't seen any of those links, and my Google-fu is failing me (spent about 10 minutes searching); would love a pointer if you have it easily at hand. (I saw the _One JavaScript_ stuff, but that is addressing a different issue, I believe.)

Separately, it's also worth noting that there's another syntax for solving this problem that I just learned about 20 minutes ago for ES6 modules. It looks like this:

```
<script type="module" src="module.js"></script>
<script nomodule src="fallback.js"></script>
```

Not sure if that's ideal for this problem, but it's another possibility. Credit to [Jake Archibald's ES6 modules article](https://jakearchibald.com/2017/es-modules-in-browsers/), which is where I learned it.

**Comment by syg** [comment]
@ljharb Yeah, versioning will be very contentious. Please save the good arguments for committee!

**Comment by ljharb** [comment]
There is a difference, though, between distinguishing parsing goals, and distinguishing spec versions.

Module and Script are fine - ES2017 vs ES2018 isn't.

**Comment by aickin** [comment]
@ljharb 

> There is a difference, though, between distinguishing parsing goals, and distinguishing spec versions.

To be clear, my last post was not saying that the existence of `nomodule` meant that version fallback should be supported natively in browsers, too; I really intended it as a totally separate comment to add to the ongoing discussion in the thread about possible syntaxes for including binary ASTs in `script` tags. I probably should have made two separate comments to make that clear.

**Comment by Yoric** [comment]
I was personally leaning towards headers, as it is pretty uniform, with a `<script type="application/binjs">` as the simple way to request a binjs code.

**Comment by littledan** [comment]
Should we do something parallel to modules here? One part is `<script type="binjs">` (usable inline?) and another part would be a `nobinjs` modifier to provide the legacy fallback, similar to `nomodule`. `Accept` headers sound like a good complementary thing as well, but it seems like it would be also useful to enable easy switching based on binjs support that *doesn't* require any logic from the web server.

**Comment by Qix-** [comment]
@Yoric that would impede adoption since you'd have to perform user agent string checking to determine if the user's browser _might_ support it.

**Comment by guybedford** [comment]
Ideally it should be possible for JS modules to load binary AST files, and for binary AST files to load normal JS as well (say for example using code from CDN would be useful - one advertising component should haven't to stop the whole codebase from using binary modules).

The plan with Web Assembly was always to use the binary header bytes which as non valid unicode to make the detection part of the binary, to be able to fit into the interop transparently. This would naturally work with this kind of a binary AST format, and would be the preferred method in my opinion.

There was an argument from the WhatWG side against this, I believe along the lines that everything loaded from `<script type="module">` should correspond to a `application/javascript` MIME type. (that is, `<script type="module">` will not load anything that does not set the valid JavaScript MIME type).

Personally I think this is the WhatWG taking their opinions about MIME here a little too far (it's fine for JS, but binary indicators could override the MIME restrictions), and that binary bytes will be an extensible route forward for the platform avoiding the complexity of splitting into careful labelling of resources or having to suddenly make the web work in terms of file extensions.

**Comment by guybedford** [comment]
I realise the above is somewhat off-topic here, if it's already discussed / answered elsewhere please let me know. Happy to move this point to a separate issue as well.

**Comment by guybedford** [comment]
For the module target, it's looking very likely at this point that Web Assembly will be determined by the `Content-Type` MIME response when using `<script type="module" src="app.wasm">`. So the detection of BinaryAST in the module target scenario could likely just extend the same technique - `<script type="module" src="app.bast">` (or whatever the extension is), where the server is returning the correct BinaryAST MIME type.
